KEP-1880 Multiple Service CIDRs: Graduate to Beta (1/2)

aojea opened the pull request:
/kind feature

Add missing criteria and close some pending work, a follow up PR will update the types to beta

- [x] Despite the tree is more optimal it is more complex than a linear search, and since the number of ServiceCIDRs is not expected to be large, use a linear search instead.
- [x] Add tests for enable and disable the feature. 
- [x] Add metrics to record the latency.
- [x] Evaluate performance
- [x] Add dual-write feature gate to avoid duplicates ClusterIP on upgrades 

Ref: https://github.com/kubernetes/enhancements/pull/4645/

```release-note
KEP-1880: Users of the new feature to add multiple service CIDR will use by default a dual-write strategy on the new ClusterIP allocators to avoid the problem of possible duplicate IPs allocated to Services when running skewed kube-apiservers using different allocators. They can opt-out of this behavior by enabled the feature gate DisableAllocatorDualWrite
```

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator_test.go:
```diff
@@ -225,12 +224,12 @@ func TestCIDRAllocateShadow(t *testing.T) {
 		t.Fatal(err)
 	}
 	// allocate one IP from the new allocator
-	err = r.Allocate(netutils.ParseIPSloppy("192.168.1.0"))
```
this was a bug since the zero address can not be allocated

aojea committed:
replace iptree on the repairip controller

aojea committed:
replace iptree on the servicecidr controller

aojea committed:
remove iptree from tree

aojea commented:
/assign @danwinship @thockin 

This remove the ip tree from the codebase

aojea commented:
/test

aojea committed:
test enable and disable ServiceCIDR feature

aojea commented:
test failure is legit, I screwed something , on the good side the testing is catching it

aojea committed:
add log to migration test

aojea committed:
add new metric to record the latency to allocate an IP address

aojea committed:
replace the iptree on the ipallocator

ServiceCIDRs are protected by finalizers and the CIDRs fields are
inmutable once set, only the readiness state impact the allocator
as it can only allocate IPs if any of the ServiceCIDR is ready.

The Add/Update events triggers a reconcilation of the current state
of the ServiceCIDR present in the informers with the existing IP
allocators.

The Delete events are handled directly to update or delete the
corresponing IP allocator.

aojea committed:
performance test to compare allocators

The current results with 100 works and 15k services on a (n2-standard-48) vCPU: 48 RAM: 192 GB are:
Old allocator:

perf_test.go:139: [RESULT] Duration 1m9.646167533s: [quantile:0.5  value:0.462886801 quantile:0.9  value:0.496662838 quantile:0.99  value:0.725845905]

New allocator:
perf_test.go:139: [RESULT] Duration 2m12.900694343s: [quantile:0.5  value:0.481814448 quantile:0.9  value:1.3867615469999999 quantile:0.99  value:1.888190671]

The new allocator has higher latency but in contrast allow to use a
larger number of services, when tested with 65k Services the old
allocator etcd crashes with storage exceeded.

The scenario is also not realistic, as a continuous and high load on
Service creation is not expected.

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -113,203 +115,228 @@ func NewMetaAllocator(
+				klog.Infof("error creating new IPAllocator for Service CIDR %s", cidr)
+				continue
+			}
+			if c.metrics {
```
can this race?

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -113,203 +115,228 @@ func NewMetaAllocator(
-			return true
+	// Deleting
+	if !serviceCIDR.DeletionTimestamp.IsZero() {
+		for _, cidr := range serviceCIDR.Spec.CIDRs {
```
ditto https://github.com/kubernetes/kubernetes/pull/122047/files#r1609614103

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -113,203 +115,228 @@ func NewMetaAllocator(
-	}()
+	// Updating
+	ready := isReady(serviceCIDR)
+	for _, cidr := range serviceCIDR.Spec.CIDRs {
```
ditto https://github.com/kubernetes/kubernetes/pull/122047/files#r1609614103

danwinship reviewed (commented):
(doesn't cover "replace the iptree on the ipallocator" yet)

aojea commented on pkg/api/servicecidr/servicecidr.go:
```diff
@@ -0,0 +1,118 @@
+)
+
+// ContainsIP return the list of ServiceCIDR that contains the IP address passed as argument
+// TODO: consolidate once danwinship implement generics to handle both net and netip types
```
ok, then will keep as it is removing the comment

aojea commented on pkg/api/servicecidr/servicecidr.go:
```diff
@@ -0,0 +1,118 @@
+
+	for _, serviceCIDR := range serviceCIDRList {
+		for _, cidr := range serviceCIDR.Spec.CIDRs {
+			if prefix, err := netip.ParsePrefix(cidr); err == nil { // it can not fail since is already validated
```
it´s also to avoid the linter 🤷

aojea commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -213,24 +207,17 @@ func (c *Controller) deleteIPAddress(obj interface{}) {
 	}
 
-	return serviceCIDRs.UnsortedList()
+	return result.UnsortedList()
```
I think that can be addressed later, I just want the diff to be small to simplify the reviews , I want it to get into the release ASAP to have enough soak time

aojea commented on pkg/api/servicecidr/servicecidr.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// ContainsAddress return the list of ServiceCIDR that contains the address passed as argument
+func ContainsAddress(serviceCIDRLister networkinglisters.ServiceCIDRLister, address netip.Addr) []*networkingv1alpha1.ServiceCIDR {
```
`ContainsNetIP` is still more confusing : /

dashpole commented:
/assign @dgrisonnet
to review the new metric

aojea committed:
implement dual write on Service ClusterIP allocator

MultiCIDRServiceAllocator implements a new ClusterIP allocator based on
IPAddress object to solve the problems and limitations caused by
existing bitmap allocators.

However, during the rollout of new versions, deployments need to support
a skew of one version between kube-apiservers. To avoid the possible
problem where there are multiple Services requests on the skewed
apiservers and that both allocate the same IP to different Services,
the new allocator will implement a dual-write strategy under the
feature gate DisableAllocatorDualWrite.

After the MultiCIDRServiceAllocator is GA, the DisableAllocatorDualWrite
can be enabled safely as all apiservers will run with the new
allocators. The graduation of DisableAllocatorDualWrite can also
be used to clean up the opaque API object that contains the old bitmaps.

If MultiCIDRServiceAllocator is enabled and DisableAllocatorDualWrite is disable
and is a new environment, there is no bitmap object created, hence, the
apiserver will initialize it to be able to write on it.

aojea committed:
repairip wait for the default servicecidr

aojea committed:
allow to shutdown quickly the apiserver

aojea commented:
linter job failure is legit

kind-ipv6 is a flake reported in https://github.com/kubernetes/kubernetes/issues/122270

seans3 commented:
/remove-sig api-machinery

aojea commented:
ping @thockin @danwinship @khenidak for reviews

thockin reviewed (commented):
minor issues on this commit

thockin commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -97,46 +117,77 @@ func NewMetaAllocator(
 }
 
-func (c *MetaAllocator) addServiceCIDR(obj interface{}) {
+func (c *MetaAllocator) enqueServiceCIDR(obj interface{}) {
```
enqueue

aojea replied:
grrrr I hate this word `queue`  😄

thockin commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -176,140 +226,135 @@ func (c *MetaAllocator) processNextItem() bool {
+				continue
+			}
+			// the readiness state of an allocator is an OR of all readiness states
+			ready := true
```
minor, but I'd do this the opposite way:

```
// readiness is the AND of several things
ready := false
if isReady(serviceCIDR) && serviceCIDR.DeletionTimestamp.IsZero() {
				ready = true
}

thockin commented on test/integration/servicecidr/feature_enable_disable.go:
```diff
@@ -0,0 +1,142 @@
+
+	// shutdown apiserver with the feature enabled
+	s2.TearDownFn()
+	// start an apiserver without the deature enabled
```
minor: "with the feature disabled"

thockin commented on pkg/registry/core/rest/storage_core.go:
```diff
@@ -351,6 +352,37 @@ func (c *Config) newServiceIPAllocators() (registries rangeRegistries, primaryCl
+		var bitmapAllocator ipallocator.Interface
+		if !utilfeature.DefaultFeatureGate.Enabled(features.DisableAllocatorDualWrite) {
+			bitmapAllocator, err = ipallocator.New(&serviceClusterIPRange, func(max int, rangeSpec string, offset int) (allocator.Interface, error) {
+				mem := allocator.NewAllocationMapWithOffset(max, rangeSpec, offset)
```
I don't love that this is an exact dup of above - pull it into a func?

aojea replied:
is not an exact dupe, it has the logic to initialze the opaque object

```
	rangeRegistry.Range = serviceClusterIPRange.String()
				if len(rangeRegistry.ResourceVersion) == 0 {
					klog.Infof("kube-apiserver started with IP allocator and dual write enabled but bitmap allocator does not exist, recreating it ...")
					err := etcd.CreateOrUpdate(rangeRegistry)
					if err != nil {
						return nil, err
					}
				}
```
another benefit is that the diffs will be cleaner once we remove the feature gates, despite now is uglier to read

thockin reviewed (commented):
Mostly minor

aojea commented on pkg/api/servicecidr/servicecidr.go:
```diff
@@ -0,0 +1,118 @@
+// covering prefix will not allocate those IPs, so a service with one of those IPs
+// can't belong to that ServiceCIDR.
+func prefixContainsIP(prefix netip.Prefix, ip netip.Addr) bool {
+	// if the IP is the network address is not contained
```
we already discussed this I can't' remember where and when, if my mind does not fail, we concluded:
- you are technically right
- the rest of the world may not be technically right and may do assumptions on these ips and not work correctly

it will be change for sure, will it be breaking? if some service implementation makes the same assumptions, I think IPVS may , we may end being technically correct but not implementable depending on project or technology.

I don't think at this point this is a problem or worth to change

thockin replied:
LOL, I do not recall that conversation, but it's plausible.  I guess we set precedent by apiserver being .1 instead of .0.

Can we document this in API comments for `ServiceCIDRSpec.CIDRs`.  E.g. "The first and last IP of each CIDR are reserved, and will not be allocated"

aojea replied:
for IPv6 the last is not reserved 🥇

aojea replied:
added an additional comment documenting the current details about allocatable IPs

aojea commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -557,60 +515,9 @@ func (r *RepairIPAddress) syncIPAddress(key string) error {
-	defer r.muTree.Unlock()
-	r.tree = tree
-	return nil
+// isIPOutOfRange returns false if the IP is not contained in any of the ServiceCIDRs
```
yes, you are right

aojea committed:
fix typo enque should be enqueue

aojea committed:
use an AND to define the allocator readiness

aojea committed:
fix wrong comment on method to detect IPs out of ranges

aojea committed:
simplify test code using the if with a short statement

aojea committed:
fix comment on integration test

dgrisonnet reviewed (approved):
The new metric looks good to me

aojea committed:
document the API with the details about the allocation CIDR ranges

thockin commented:
Thanks!

/lgtm
/approve

aojea committed:
generated

aojea commented:
> Thanks!
> 
> /lgtm /approve

failed the generated code , added in the last commit

thockin reviewed (commented):
Thanks!

/lgtm
/approve

pacoxu commented:
/hold 
for ci failure: `Please run 'hack/update-openapi-spec.sh'`

aojea committed:
update openapi

aojea commented:
oh, I need to do make update, will push a new commit, @pacoxu do you 

> /hold for ci failure: `Please run 'hack/update-openapi-spec.sh'`

/hold cancel

I forget to update the openapi, added it in the last commit

@pacoxu do you mind to lgtm again? the code is the same just with this missing generated code

pacoxu commented:
/lgtm
